server/sonar-server/src/test/java/org/sonar/server/computation/issue/IssueComputationTest.java

```diff
@@ -0,0 +1,152 @@
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  // inputs
+  RuleCache ruleCache = mock(RuleCache.class);
+  SourceLinesCache lineCache = mock(SourceLinesCache.class);
+  ScmAccountCache scmAccountCache = mock(ScmAccountCache.class);
+  DefaultIssue issue = new DefaultIssue().setRuleKey(RULE_KEY).setKey("ISSUE_A");
+  RuleDto rule = new RuleDto().setRepositoryKey(RULE_KEY.repository()).setRuleKey(RULE_KEY.rule());
+
+  // output
+  FinalIssues finalIssues;
+
+  @Before
+  public void setUp() throws IOException {
+    when(ruleCache.getNullable(RULE_KEY)).thenReturn(rule);
+    finalIssues = new FinalIssues(temp.newFile(), System2.INSTANCE);
+  }
+
+  @Test
+  public void store_issues_on_disk() throws Exception {
+    process();
+
+    assertThat(Iterators.getOnlyElement(finalIssues.traverse()).key()).isEqualTo("ISSUE_A");
+  }
+
+  @Test
+  public void copy_rule_tags_on_new_issues() throws Exception {
+    issue.setNew(true);
+    rule.setTags(ImmutableSet.of("bug", "performance"));
+    rule.setSystemTags(ImmutableSet.of("blocker"));
+
+    process();
+
+    assertThat(issue.tags()).containsOnly("blocker", "bug", "performance");
+  }
+
+  @Test
+  public void do_not_copy_rule_tags_on_existing_issues() throws Exception {
+    issue.setNew(false);
+    rule.setTags(ImmutableSet.of("bug", "performance"));
+    rule.setSystemTags(ImmutableSet.of("blocker"));
+
+    process();
+
+    assertThat(issue.tags()).isEmpty();
+  }
+
+  @Test
+  public void guess_author_of_new_issues() throws Exception {
+    issue.setNew(true);
+    issue.setLine(3);
+    when(lineCache.lineAuthor(3)).thenReturn("charlie");
+
+    process();
+
+    assertThat(issue.authorLogin()).isEqualTo("charlie");
+  }
+
+  @Test
+  public void do_not_fail_if_missing_author_for_new_issues() throws Exception {
+    issue.setNew(true);
+    issue.setLine(3);
+    when(lineCache.lineAuthor(3)).thenReturn(null);
+
+    process();
+
+    assertThat(issue.authorLogin()).isNull();
+  }
+
+  @Test
+  public void do_not_guess_author_of_existing_issues() throws Exception {
+    issue.setNew(false);
+    issue.setLine(3);
+    when(lineCache.lineAuthor(3)).thenReturn("charlie");
+
+    process();
+
+    assertThat(issue.authorLogin()).isNull();
+  }
+
+  @Test
+  public void auto_assign_new_issues() throws Exception {
+    issue.setNew(true);
+    issue.setAuthorLogin("charlie");
+    when(scmAccountCache.getNullable("charlie")).thenReturn("char.lie");
+
+    process();
+
+    assertThat(issue.assignee()).isEqualTo("char.lie");
+  }
+
+  @Test
+  public void do_not_auto_assign_existing_issues() throws Exception {
+    issue.setNew(false);
+    issue.setAuthorLogin("charlie");
+    when(scmAccountCache.getNullable("charlie")).thenReturn("char.lie");
+
+    process();
+
+    assertThat(issue.assignee()).isNull();
+  }
+
+  private void process() {
+    IssueComputation computation = new IssueComputation(ruleCache, lineCache, scmAccountCache, finalIssues);
```

Comment:
same pattern as most tests, easy to see where the IssueComputation is instanciated. process should just process